plugin/documents.py

```diff
@@ -216,6 +265,95 @@ def on_hover(self, point: int, hover_zone: int) -> None:
             return
         self.view.run_command("lsp_hover", {"point": point})
 
+    # --- textDocument/signatureHelp -----------------------------------------------------------------------------------
+
+    def _do_signature_help(self) -> None:
+        # NOTE: We take the beginning of the region to check the previous char (see last_char variable). This is for
+        # when a language server inserts a snippet completion.
+        pos = self._stored_region.a
+        if pos == -1:
+            return
+        if not self.view.match_selector(pos, self.view.settings().get("auto_complete_selector") or ""):  # ???
+            return
+        provider = None  # type: Optional[Dict[str, Any]]
+        session = None  # type: Optional[Session]
+        for sv in self.session_views_async():
+            provider = sv.session.get_capability("signatureHelpProvider")
+            if provider:
+                session = sv.session
+                break
+        if not provider or not session:
+            return
+        triggers = provider.get("triggerCharacters") or []
+        if not triggers:
+            return
+        last_char = self.view.substr(pos - 1)
+        if last_char in triggers:
+            self.purge_changes_async()
+            params = text_document_position_params(self.view, pos)
+            session.send_request(Request.signatureHelp(params), lambda resp: self._on_signature_help(resp, pos))
+        elif self.view.is_popup_visible():
+            if last_char.isspace():
+                # Peek behind to find the last non-whitespace character.
+                last_non_white_space_position = self.view.find_by_class(pos, False, ~0)
+                last_char = self.view.substr(last_non_white_space_position - 1)
+            if last_char not in triggers:
+                # TODO: Refactor popup usage to a common class. We now have sigHelp, completionDocs, hover, and diags
+                # all using a popup. Most of these systems assume they have exclusive access to a popup, while in
+                # reality there is only one popup per view.
+                self.view.hide_popup()
+
+    def _on_signature_help(self, response: Optional[Dict], point: int) -> None:
+        region = self._get_current_region_async()
+        if region is None:
+            return
+        if point != region.b:
+            return
+        self._sighelp = create_signature_help(response)
+        if self._sighelp:
+            content = self._sighelp.build_popup_content(self._sighelp_renderer)
+
+            def render_sighelp_on_main_thread() -> None:
+                if self.view.is_popup_visible():
+                    self._update_popup(content)
+                else:
+                    self._show_popup(content, point)
+
+            sublime.set_timeout(render_sighelp_on_main_thread)
+
+    def _show_popup(self, content: str, point: int) -> None:
```

Comment:
Now that various functionality is within the same class it would probably be better to make method names more unique like `_show_signature_popup`, `_update_signature_popup`, `on_hide_signature_popup`...